pylot/drivers/carla_gnss_driver_operator.py

```diff
@@ -0,0 +1,104 @@
+    to the vehicle, registers callback functions to retrieve the GNSS
+    measurements and publishes it to downstream operators.
+
+    Args:
+        ground_vehicle_id_stream (:py:class:`erdos.ReadStream`): Stream on
+            which the operator receives the id of the ego vehicle. It uses this
+            id to get a Carla handle to the vehicle.
+        gnss_stream (:py:class:`erdos.WriteStream`): Stream on which the
+            operator sends GNSS info.
+        gnss_setup (:py:class:`pylot.drivers.sensor_setup.GNSSSetup`):
+            Setup of the GNSS sensor.
+        flags (absl.flags): Object to be used to access absl flags.
+    """
+
+    def __init__(self, ground_vehicle_id_stream, gnss_stream, gnss_setup,
+                 flags):
+        # Save the streams.
+        self._vehicle_id_stream = ground_vehicle_id_stream
+        self._gnss_stream = gnss_stream
+
+        # Save the flags and initialize logging.
+        self._flags = flags
+        self._logger = erdos.utils.setup_logging(self.config.name,
+                                                 self.config.log_file_name)
+
+        # Save the setup, the vehicle and the sensor.
+        self._gnss_setup = gnss_setup
+        self._vehicle = None
+        self._gnss = None
+        self._lock = threading.Lock()
+
+    @staticmethod
+    def connect(ground_vehicle_id_stream):
+        gnss_stream = erdos.WriteStream()
+        return [gnss_stream]
+
+    def process_gnss(self, gnss_msg):
+        """ Invoked when a GNSS message is received from the simulator.
+
+        Sends GNSS measurements to downstream operators.
+
+        Args:
+            gnss_msg: carla.GnssMeasurement
+        """
+        game_time = int(gnss_msg.timestamp * 1000)
+        timestamp = erdos.Timestamp(coordinates=[game_time])
+        watermark_msg = erdos.WatermarkMessage(timestamp)
+        with erdos.profile(self.config.name + '.process_gnss',
+                           self,
+                           event_data={'timestamp': str(timestamp)}):
+            with self._lock:
+                msg = GNSSMessage(
+                    timestamp,
+                    Transform.from_carla_transform(gnss_msg.transform),
+                    gnss_msg.altitude, gnss_msg.latitude, gnss_msg.longitude)
+                self._gnss_stream.send(msg)
+                self._gnss_stream.send(watermark_msg)
+
+    def run(self):
+        # Read the vehicle ID from the vehicle ID stream.
+        vehicle_id = self._vehicle_id_stream.read().data
+        self._logger.debug(
+            "The CarlaGNSSDriverOperator received the vehicle id: {}".format(
+                vehicle_id))
+
+        # Connect to the world.
+        _, world = get_world(self._flags.carla_host, self._flags.carla_port,
+                             self._flags.carla_timeout)
+        if world is None:
+            raise ValueError("There was an issue connecting to the simulator.")
```

Comment:
No need to raise an Error here because get_world is going to raise an exception anyways if it can't get the world.